### PR TITLE
Add more type annotations

### PR DESCRIPTION
**Pull request by and-semakin**
In this PR I enabled strict `mypy` settings & fixed some reasonable warnings. Also added `mypy` and `flake8` linters to CI to automatically check all the commits.

mypy settings I used:
```
disallow_untyped_calls = True
disallow_untyped_defs = True
disallow_incomplete_defs = True
disallow_untyped_decorators = True
strict_optional = True
warn_redundant_casts = True
warn_no_return = True
warn_return_any = True
warn_unreachable = True
```

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -24,6 +24,8 @@ install:
 
 script:
   - poetry run py.test tests/
+  - poetry run mypy rsa/
```
Comment (and-semakin) [inline .travis.yml]: Just found that `mypy` is already running in CI as a test case in `pytest`.

Do I need to add `flake8` as a test also or can I just drop the `mypy` test?

Comment (sybrenstuvel) [reply .travis.yml]: Please also add a flake8 test. That way it's possible to just run `py.test` and have everything tested in one go.

**Comment by sybrenstuvel** [comment]
Thanks!

File: .travis.yml (hunk 1 above)
Comment (and-semakin) [reply .travis.yml]: I would like to make it in a separate PR to keep it focused. I will revert my current `.travis.yml` changes.

**Comment by and-semakin** [comment]
By the way it seems that Travis CI is set in some wrong way in this repo. I would expect it to block my PR if something is broken in tests. Now I see errors only in my own fork but not here.

**Comment by and-semakin** [comment]
@sybrenstuvel tell me if I need to improve something

**Pull request merged**